File: NRDB/CompareDecksList.m
//
//  CompareDecksList.m
//  NRDB
//
//  Created by Gereon Steffens on 15.07.14.
//  Copyright (c) 2014 Gereon Steffens. All rights reserved.
//

#import <SDCAlertView.h>

#import "CompareDecksList.h"
#import "DeckCell.h"
#import "Deck.h"
#import "DeckDiffViewController.h"
#import "DeckManager.h"

@interface CompareDecksList ()

@property NSMutableArray* decksToDiff;
@property NSMutableArray* names;

@property UIBarButtonItem* diffButton;
@property UIToolbar* toolbar;
@property UILabel* footerLabel;

@property NRRole selectedRole;

@end

@implementation CompareDecksList

-(void) viewDidLoad
{
    [super viewDidLoad];
    self.decksToDiff = [NSMutableArray array];
    self.names = [NSMutableArray array];
    self.selectedRole = NRRoleNone;
    
    self.diffButton = [[UIBarButtonItem alloc] initWithTitle:l10n(@"Compare") style:UIBarButtonItemStylePlain target:self action:@selector(diffDecks:)];
    self.diffButton.enabled = NO;
    
    UINavigationItem* topItem = self.navigationController.navigationBar.topItem;
    topItem.rightBarButtonItems = @[ self.diffButton ];

    // add toolbar as footer
    self.toolbar = [[UIToolbar alloc] initWithFrame:CGRectMake(0, 728, 703, 40)];
    self.footerLabel = [[UILabel alloc] initWithFrame:CGRectMake(10, 0, 683, 40)];
    self.footerLabel.font = [UIFont systemFontOfSize:15];
    [self.toolbar addSubview:self.footerLabel];
    self.footerLabel.text = l10n(@"Select two decks to compare them");
    self.footerLabel.userInteractionEnabled = YES;
    UITapGestureRecognizer* tap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(footerClicked:)];
    [self.footerLabel addGestureRecognizer:tap];
    [self.view addSubview:self.toolbar];
    
    // adjust tableview
    UIEdgeInsets insets = self.tableView.contentInset;
    insets.bottom = 40;
    self.tableView.contentInset = insets;
}

-(void) diffDecks:(id)sender
{
    NSAssert(self.decksToDiff.count == 2, @"count must be 2");
    
    Deck* deck1 = [DeckManager loadDeckFromPath:self.decksToDiff[0]];
    Deck* deck2 = [DeckManager loadDeckFromPath:self.decksToDiff[1]];
    
    if (deck1.role != deck2.role)
    {
        [SDCAlertView alertWithTitle:nil message:l10n(@"Both decks must be for the same side.") buttons:@[ l10n(@"OK")] ];
        return;
    }
    
    [DeckDiffViewController showForDecks:deck1 deck2:deck2 inViewController:self];
}

-(void) footerClicked:(id)sender
{
    if (self.decksToDiff.count == 2)
    {
        [self diffDecks:nil];
    }
}

#pragma mark table view

-(UITableViewCell*) tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    DeckCell* cell = (DeckCell*)[super tableView:tableView cellForRowAtIndexPath:indexPath];
    
    NSArray* decks = self.decks[indexPath.section];
    Deck* deck = decks[indexPath.row];
    
    cell.accessoryType = UITableViewCellAccessoryNone;
    if ([self.decksToDiff containsObject:deck.filename])
    {
        cell.accessoryType = UITableViewCellAccessoryCheckmark;
    }
    
    return cell;
}

-(void) tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    NSArray* decks = self.decks[indexPath.section];
    Deck* deck = decks[indexPath.row];
    
    if (self.selectedRole != deck.role)
    {
        [self.decksToDiff removeAllObjects];
        [self.names removeAllObjects];
    }
    self.selectedRole = deck.role;
    
    if ([self.decksToDiff containsObject:deck.filename])
    {
        [self.decksToDiff removeObject:deck.filename];
        [self.names removeObject:deck.name];
    }
    else
    {
        [self.decksToDiff addObject:deck.filename];
        [self.names addObject:deck.name];
    }
    
    while (self.decksToDiff.count > 2)
    {
        [self.decksToDiff removeObjectAtIndex:0];
        [self.names removeObjectAtIndex:0];
    }
    
    self.diffButton.enabled = self.decksToDiff.count == 2;
    
    switch (self.decksToDiff.count)
    {
        case 0:
            self.footerLabel.text = l10n(@"Select two decks to compare them");
            self.footerLabel.textColor = [UIColor blackColor];
            break;
        case 1:
            self.footerLabel.text = [NSString stringWithFormat:l10n(@"Selected ‘%@’, select one more to compare"), self.names[0]];
            self.footerLabel.textColor = [UIColor blackColor];
            break;
        case 2:
            self.footerLabel.text = [NSString stringWithFormat:l10n(@"Selected ‘%@’ and ‘%@’, tap to compare"), self.names[0], self.names[1]];
            self.footerLabel.textColor = UIColorFromRGB(0x007aff);
            break;
    }
    
    [self.tableView reloadData];
}

@end

File: NetDeck/CompareDecksList.swift
//
//  CompareDecksList.swift
//  NetDeck
//
//  Created by Gereon Steffens on 03.01.17.
//  Copyright © 2021 Gereon Steffens. All rights reserved.
//

import UIKit

class CompareDecksList: DecksViewController {
    
    private var decksToDiff = [String]()
    private var names = [String]()
    private var diffButton: UIBarButtonItem!
    private var footerButton: UIBarButtonItem!
    private var selectedRole = Role.none
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.diffButton = UIBarButtonItem(title: "Compare".localized(), style: .plain, target: self, action: #selector(self.diffDecks(_:)))
        self.diffButton.isEnabled = false
        
        self.navigationItem.rightBarButtonItem = self.diffButton
        
        let title = "Select two decks to compare them".localized()
        self.footerButton = UIBarButtonItem(title: title, style: .plain, target: self, action: #selector(self.footerClicked(_:)))
        self.footerButton.tintColor = .label
        let fontName = [ NSAttributedString.Key.font: UIFont.systemFont(ofSize: 15)]
        self.footerButton.setTitleTextAttributes(fontName, for: .normal)
        
        self.toolBar.items = [ self.footerButton ]
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)

        var scrollTo: IndexPath?
        if self.decks.count > 0 && self.decks[0].count > 0 {
            scrollTo = IndexPath(row: 0, section: 0)
        } else if self.decks.count > 1 && self.decks[1].count > 0 {
            scrollTo = IndexPath(row: 0, section: 1)
        }

        if let top = scrollTo {
            self.tableView.scrollToRow(at: top, at: .top, animated: false)
        }
    }

    @objc func diffDecks(_ sender: UIBarButtonItem) {
        if self.popup != nil {
            return self.dismissPopup()
        }
        
        assert(self.decksToDiff.count == 2, "count must be 2")
        
        let d1 = DeckManager.loadDeckFromPath(self.decksToDiff[0])
        let d2 = DeckManager.loadDeckFromPath(self.decksToDiff[1])
        
        guard let deck1 = d1, let deck2 = d2 else {
            return
        }
        
        if deck1.role != deck2.role {
            UIAlertController.alert(withTitle: nil, message: "Both decks must be for the same side.".localized(), button: "OK".localized())
            return
        }
        
        DeckDiffViewController.showForDecks(deck1, deck2: deck2, inViewController: self)
    }
    
    @objc func footerClicked(_ sender: UIBarButtonItem) {
        if self.decksToDiff.count == 2 {
            self.diffDecks(sender)
        }
    }
    
    // MARK: - table view
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = super.tableView(tableView, cellForRowAt: indexPath) as! DeckCell
        
        let deck = self.decks[indexPath.section][indexPath.row]
        
        cell.infoButton?.isHidden = true
        cell.accessoryType = .none
        
        if self.decksToDiff.contains(deck.filename!) {
            cell.infoButton?.isHidden = false
            cell.infoButton?.isUserInteractionEnabled = false
            
            let img = UIImage(named: "888-checkmark-selected")?.withRenderingMode(.alwaysTemplate)
            cell.infoButton?.setImage(img, for: .normal)
        }
        
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        let deck = self.decks[indexPath.section][indexPath.row]
        let filename = deck.filename!
        
        if self.selectedRole != deck.role {
            self.decksToDiff.removeAll()
            self.names.removeAll()
        }
        self.selectedRole = deck.role
        
        if let index = self.decksToDiff.firstIndex(of: filename) {
            self.decksToDiff.remove(at: index)
            
            for i in 0 ..< self.names.count {
                if self.names[i] == deck.name {
                    self.names.remove(at: i)
                    break
                }
            }
        } else {
            self.decksToDiff.append(filename)
            self.names.append(deck.name)
        }
        
        while self.decksToDiff.count > 2 {
            self.decksToDiff.remove(at: 0)
            self.names.remove(at: 0)
        }
        
        assert(self.decksToDiff.count == self.names.count, "count mismatch")
        self.diffButton.isEnabled = self.decksToDiff.count == 2
        
        switch self.decksToDiff.count {
        case 1:
            self.footerButton.title = String(format: "Selected ‘%@’, select one more to compare".localized(), self.names[0])
            self.footerButton.tintColor = .label
        case 2:
            self.footerButton.title = String(format: "Selected ‘%@’ and ‘%@’, tap to compare".localized(), self.names[0], self.names[1])
            self.footerButton.tintColor = self.diffButton.tintColor
        default:
            self.footerButton.title = "Select two decks to compare them".localized()
            self.footerButton.tintColor = .label
        }
        
        self.tableView.reloadData()
    }
}

// MARK: - Empty state
extension CompareDecksList {
    
    func buttonTitle(forEmptyDataSet scrollView: UIScrollView!, for state: UIControl.State) -> NSAttributedString! {
        return nil
    }
    
}
